src/v2/auth.rs

```diff
@@ -1,23 +1,161 @@
+    pub(crate) fn from_www_authentication_header(header_value: HeaderValue) -> Result<Self> {
+        let header = String::from_utf8(header_value.as_bytes().to_vec())?;
+
+        // This regex will result in multiple captures which will contain one key-value pair each.
+        // The first capture will be the only one with the "method" group set.
+        let re = regex::Regex::new(
+            r#"(?x)\s*
+            ((?P<method>[A-Z][a-z]+)\s*)?
+            (
+                \s*
+                    (?P<key>[a-z]+)
+                \s*
+                    =
+                \s*
+                    "(?P<value>[^"]+)"
+                \s*
+            )
+        "#,
+        )?;
+        let captures = re.captures_iter(&header).collect::<Vec<_>>();
+
+        let method = captures
+            .get(0)
+            .ok_or_else(|| {
+                Error::from(format!("regex '{}' didn't match '{}'", re.as_str(), header))
+            })?
+            .name("method")
+            .ok_or_else(|| Error::from(format!("method not found in {}", header)))?
+            .as_str();
+
+        let serialized_content = format!(
+            r#"{{ "{}": {{ {} }} }}"#,
+            method,
+            captures
+                .iter()
+                .filter_map(|capture| {
+                    match (
+                        capture.name("key").map(|n| n.as_str().to_string()),
+                        capture.name("value").map(|n| n.as_str().to_string()),
+                    ) {
+                        (Some(key), Some(value)) => Some(format!(r#""{}": "{}""#, key, value)),
+                        _ => None,
+                    }
+                })
+                .collect::<Vec<_>>()
+                .join(", "),
+        );
+
+        // Deserialize the content
+        let mut unsupported_keys = std::collections::HashSet::new();
+        let content: WwwAuthenticateHeaderContent = serde_ignored::deserialize(
+            &mut serde_json::Deserializer::from_str(&serialized_content),
+            |path| {
+                unsupported_keys.insert(path.to_string());
+            },
+        )?;
+
+        if !unsupported_keys.is_empty() {
+            warn!("unsupported keys remaining {:#?}", unsupported_keys);
+        }
+
+        Ok(content)
     }
 }
 
+/// Structured content for the Bearer authentication response header.
+#[derive(Debug, Default, PartialEq, Eq, Deserialize)]
+pub(crate) struct WwwAuthenticateHeaderContentBearer {
+    realm: String,
+    service: Option<String>,
+    scope: Option<String>,
+}
+
+impl WwwAuthenticateHeaderContentBearer {
+    fn auth_ep(&self, scopes: &[&str]) -> String {
+        let service = self
+            .service
+            .as_ref()
+            .map(|sv| format!("?service={}", sv))
+            .unwrap_or_default();
+
+        let scope = scopes.iter().enumerate().fold(
+            if scopes.is_empty() {
+                ""
+            } else if service.is_empty() {
+                "?"
```

Comment:
I have the feeling this possibly does not belong to the folding accumulator. I would have expected this one on its own `separator`, outside of the iterator and directly plugged into the format string. What do you think?